### PR TITLE
Removed deprecated constructors and methods from Thing API

### PR DESCRIPTION
**Pull request by cweitkamp**
- Removed deprecated constructors and methods from Thing API
- Replaced deprecated constructors by related builder where possible

Relates to #1408

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

---

Depends on https://github.com/openhab/openhab-addons/pull/7352
Depends on https://github.com/openhab/openhab-addons/issues/7354
Depends on https://github.com/openhab/openhab-addons/issues/7402
Depends on https://github.com/openhab/org.openhab.binding.zwave/issues/1339

### CODE REVIEW

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/i18n/ChannelTypeI18nLocalizationService.java [hunk 1]
```diff
@@ -160,11 +160,7 @@ public ChannelType createLocalizedChannelType(Bundle bundle, ChannelType channel
                 }
                 return triggerBuilder.build();
             default:
-                return new ChannelType(channelTypeUID, channelType.isAdvanced(), channelType.getItemType(),
-                        channelType.getKind(), label == null ? defaultLabel : label, description,
-                        channelType.getCategory(), channelType.getTags(), channelType.getState(),
-                        channelType.getEvent(), channelType.getConfigDescriptionURI(),
-                        channelType.getAutoUpdatePolicy());
+                throw new IllegalArgumentException("Kind must not be null or aempty!");
```
Comment (cweitkamp) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/i18n/ChannelTypeI18nLocalizationService.java:163]:
```suggestion
                throw new IllegalArgumentException("Kind must not be null or empty!");
```

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/type/StateChannelTypeBuilderImpl.java [hunk 2]
```diff
@@ -65,15 +64,15 @@ public StateChannelTypeBuilder withCommandDescription(@Nullable CommandDescripti
         return this;
     }
 
+    @SuppressWarnings("deprecation")
     @Override
     public ChannelType build() {
         if (stateDescription != null) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/type/StateChannelTypeBuilderImpl.java:70]: Is this `if` really necessary?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import java.util.Set;
 
 import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.Nullable;
```
Comment (cweitkamp) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:21]: FTR: I am working on nullness annotations for this class.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java [hunk 4]
```diff
@@ -41,59 +42,35 @@
...
+                configDescriptionURI, null);
     }
 
     /**
+     * Creates a new instance of this class with the specified parameters.
+     *
      * @deprecated Use the {@link ChannelTypeBuilder} instead.
      */
     @Deprecated
-    public ChannelType(ChannelTypeUID uid, boolean advanced, String itemType, ChannelKind kind, String label,
```
Comment (wborn) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: Looks like there are still add-ons using these deprecated constructors. I made  local add-ons port build and it failed on the Homematic add-on:

```
HomematicTypeGeneratorImpl.java:[306,14266] The constructor ChannelType(ChannelTypeUID, boolean, String, ChannelKind, String, String, String, null, StateDescription, EventDescription, URI) is undefined
````

So if we'd merge this PR it would break the add-ons port. Unless the deprecations are fixed on the 2.5.x. branch.

Comment (MHerbst) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: I am currently working on a PR for the Homematic binding that will replace all deprecated constructors, methods etc. Will try to push it today.

Comment (wborn) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: That will certainly help @MHerbst! :+1:

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: @MHerbst Thank you. I will crawl the rest of the add-ons and replace the remaining.

@wborn How can I manage a local add-ons port for testing?

Comment (wborn) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: I use this [port-addons.sh](https://gist.github.com/wborn/0e3cd275be724e827ef88bb04b7eb12a) script based on the script used by Jenkins in the add-ons port build.

Comment (wborn) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: So you run the script on a 2.5.x based branch.

If you want to do the same build as Jenkins you checkout 2.5.x in that dir. E.g. to see if there are any openhab-core changes breaking the add-ons port build.

Sometimes you also want to checkout your own 2.5.x branch to see if you can keep 2.5.x compatible with breaking changes OH 3.0. I do that for instance when adding nullness annotations. You might also want to do that after fixing deprecations before submitting the PR.

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: My PR to fix add-ons build: https://github.com/openhab/openhab-addons/pull/7354/

**Comment by cweitkamp** [comment]
I squashed my commits and rebased to resolve the conflicts.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java (hunk 4 above)
Comment (wborn) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/type/ChannelType.java:64]: Thank you both for fixing the deprecations!
The add-ons port now again builds for me with this PR locally. :+1:

File: bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/ThingUIDTest.java [hunk 5]
```diff
@@ -26,19 +26,19 @@ public void testThreeSegments() {
...
-    @Test
-    public void testTwoSegments() {
-        ThingUID t = new ThingUID("fake", "gaga");
-
-        assertNull(t.getThingTypeId());
-        assertEquals("gaga", t.getId());
-        assertEquals("fake::gaga", t.getAsString());
-    }
+    // @Test
+    // public void testTwoSegments() {
```
Comment (wborn) [inline bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/ThingUIDTest.java:35]: Why was it commented?
Can it be uncommented, fixed or removed instead?

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/ThingUID.java [hunk 6]
```diff
@@ -135,29 +135,11 @@ public ThingUID(String... segments) {
...
-        } else {
-            return thingType;
-        }
-    }
-
     /**
      * Returns the thing type uid.
      *
      * @return thing type uid
-     * @deprecated use {@link Thing#getThingTypeUID()} instead.
```
Comment (wborn) [inline bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/ThingUID.java:158]: It's probably deprecated because this method always creates a new `ThingTypeUID`, whereas `Thing.getThingTypeUID()` returns an existing object. In a binding you typically already have a `Thing` object, so it will save CPU cycles and garbage (+ collection) by calling `Thing.getThingTypeUID()` and reusing the existing objects where possible.

File: bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/ThingUIDTest.java (hunk 5 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/ThingUIDTest.java]: I readed and changed it.

File: bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/ThingUID.java (hunk 6 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/ThingUID.java:158]: true ... and it can return `null` which` breaks other stuff like discovery ... (see e.g. https://github.com/openhab/openhab-core/pull/1414/commits/2d88d587e05d793b755aee4abfa349c8ce7803df).

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/ThingUID.java:158]: I finally found a way to remove it.

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/ThingUID.java:158]: Some more things to solve for the add-ons build: https://github.com/openhab/openhab-addons/issues/7402 and https://github.com/openhab/openhab-addons/pull/7403.

File: bundles/org.openhab.core.config.discovery/src/main/java/org/openhab/core/config/discovery/internal/DiscoveryResultImpl.java [hunk 7]
```diff
@@ -45,31 +48,6 @@
     DiscoveryResultImpl() {
     }
 
-    /**
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.discovery/src/main/java/org/openhab/core/config/discovery/internal/DiscoveryResultImpl.java:48]: Moved into separate PR (#1424) but kind of required for this PR.

**Comment by cweitkamp** [comment]
@openhab/core-maintainers How should we continue with this PR? AFAIK there still is an conflict on the Velux binding. Should we merge without waiting for it to be solved and exclude it from build if add-ons port?

**Comment by wborn** [comment]
If it is just that add-on I think it's fine to remove it from OH3 builds for now. These deprecations should be fixed anyway in OH3 and this PR has been waiting for months. It will also fix many deprecations in the [list](https://next.openhab.org/javadoc/latest/deprecated-list.html).

**Comment by kaikreuzer** [comment]
Fully agree - in two weeks we will anyhow have the add-ons being moved to master and than there'll be more incentives to fix the bindings for OH3.

**Comment by cweitkamp** [comment]
:+1: I will try to resolve the conflicts tomorrow.

**Comment by cweitkamp** [comment]
Done.

**Review by wborn — APPROVED** [review]
Thanks! It nicely cleans up a lot of deprecations and the null analysis improvements are always welcome too. :-)

**Comment by wborn** [comment]
Looks like the new annotations have uncovered some potential NPEs as the [add-ons port build](https://ci.openhab.org/job/openHAB-Addons-port/263/) is now broken due to add-on(s) using `@Nullable` fields in discovery services @cweitkamp.

**Comment by cweitkamp** [comment]
Looks like the build stopped on a new binding. I ran a local add-ons port build. Maybe it was not on latest code based on 2.5.x branch. I will have a look later.

**Comment by cweitkamp** [comment]
I submitted a fix for the annotation warnings (https://github.com/openhab/openhab-addons/pull/8473).
